File: payment-service/src/main/java/com/ecommerce/payment/service/PaymentService.java
package com.ecommerce.payment.service;

import com.ecommerce.payment.dto.PaymentRequest;
import com.ecommerce.payment.dto.PaymentResponse;
import com.ecommerce.payment.model.Payment;
import com.ecommerce.payment.model.PaymentMethod;
import com.ecommerce.payment.model.PaymentStatus;
import com.ecommerce.payment.repository.PaymentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService {
    
    private final PaymentRepository paymentRepository;
    private final Random random = new Random();
    
    @Transactional
    public PaymentResponse processPayment(PaymentRequest request) {
        log.info("Processing payment for order: {} with amount: {}", 
                request.getOrderId(), request.getAmount());
        
        // Check if payment already exists for this order
        if (paymentRepository.findByOrderId(request.getOrderId()).isPresent()) {
            throw new RuntimeException("Payment already exists for order: " + request.getOrderId());
        }
        
        // Create payment entity
        Payment payment = new Payment();
        payment.setOrderId(request.getOrderId());
        payment.setUserId(request.getUserId());
        payment.setAmount(request.getAmount());
        payment.setPaymentMethod(PaymentMethod.valueOf(request.getPaymentMethod()));
        payment.setStatus(PaymentStatus.PROCESSING);
        
        Payment savedPayment = paymentRepository.save(payment);
        log.info("Payment created with ID: {}", savedPayment.getId());
        
        // Simulate payment processing with payment gateway
        boolean paymentSuccess = simulatePaymentGateway();
        
        if (paymentSuccess) {
            savedPayment.setStatus(PaymentStatus.COMPLETED);
            savedPayment.setProcessedAt(LocalDateTime.now());
            savedPayment.setPaymentGatewayResponse("Payment processed successfully");
            paymentRepository.save(savedPayment);
            
            log.info("Payment completed successfully: {}", savedPayment.getTransactionId());
            
            return mapToResponse(savedPayment, "Payment completed successfully");
        } else {
            savedPayment.setStatus(PaymentStatus.FAILED);
            savedPayment.setProcessedAt(LocalDateTime.now());
            savedPayment.setPaymentGatewayResponse("Payment failed - Insufficient funds or gateway error");
            paymentRepository.save(savedPayment);
            
            log.error("Payment failed for order: {}", request.getOrderId());
            
            throw new RuntimeException("Payment processing failed");
        }
    }
    
    public PaymentResponse getPaymentById(Long id) {
        log.info("Fetching payment by ID: {}", id);
        
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Payment not found with ID: " + id));
        
        return mapToResponse(payment, null);
    }
    
    public PaymentResponse getPaymentByOrderId(Long orderId) {
        log.info("Fetching payment by order ID: {}", orderId);
        
        Payment payment = paymentRepository.findByOrderId(orderId)
                .orElseThrow(() -> new RuntimeException("Payment not found for order: " + orderId));
        
        return mapToResponse(payment, null);
    }
    
    public PaymentResponse getPaymentByTransactionId(String transactionId) {
        log.info("Fetching payment by transaction ID: {}", transactionId);
        
        Payment payment = paymentRepository.findByTransactionId(transactionId)
                .orElseThrow(() -> new RuntimeException("Payment not found with transaction ID: " + transactionId));
        
        return mapToResponse(payment, null);
    }
    
    public List<PaymentResponse> getAllPayments() {
        log.info("Fetching all payments");
        
        return paymentRepository.findAll().stream()
                .map(payment -> mapToResponse(payment, null))
                .collect(Collectors.toList());
    }
    
    public List<PaymentResponse> getPaymentsByUserId(Long userId) {
        log.info("Fetching payments for user: {}", userId);
        
        return paymentRepository.findByUserId(userId).stream()
                .map(payment -> mapToResponse(payment, null))
                .collect(Collectors.toList());
    }
    
    @Transactional
    public PaymentResponse refundPayment(Long id) {
        log.info("Processing refund for payment: {}", id);
        
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Payment not found with ID: " + id));
        
        if (payment.getStatus() != PaymentStatus.COMPLETED) {
            throw new RuntimeException("Only completed payments can be refunded");
        }
        
        payment.setStatus(PaymentStatus.REFUNDED);
        payment.setProcessedAt(LocalDateTime.now());
        payment.setPaymentGatewayResponse("Payment refunded successfully");
        
        Payment refundedPayment = paymentRepository.save(payment);
        log.info("Payment refunded successfully: {}", refundedPayment.getTransactionId());
        
        return mapToResponse(refundedPayment, "Payment refunded successfully");
    }
    
    // Simulate payment gateway processing (90% success rate)
    private boolean simulatePaymentGateway() {
        try {
            // Simulate processing delay
            Thread.sleep(1000);
            
            // 90% success rate
            return random.nextInt(10) < 9;
        } catch (InterruptedException e) {
            log.error("Payment processing interrupted", e);
            return false;
        }
    }
    
    private PaymentResponse mapToResponse(Payment payment, String message) {
        return PaymentResponse.builder()
                .id(payment.getId())
                .orderId(payment.getOrderId())
                .userId(payment.getUserId())
                .amount(payment.getAmount())
                .status(payment.getStatus())
                .paymentMethod(payment.getPaymentMethod())
                .transactionId(payment.getTransactionId())
                .message(message != null ? message : payment.getPaymentGatewayResponse())
                .createdAt(payment.getCreatedAt())
                .processedAt(payment.getProcessedAt())
                .build();
    }
}

File: backend/payment-service/src/services/paymentService.js
const Payment = require('../models/Payment');
const { v4: uuidv4 } = require('uuid');

class PaymentService {
  // Process payment (called by Order Service)
  async processPayment(paymentData) {
    try {
      const { orderId, amount, paymentMethod } = paymentData;

      // Create payment record
      const payment = new Payment({
        orderId,
        amount,
        paymentMethod,
        status: 'PROCESSING'
      });

      await payment.save();

      // Simulate payment processing
      // In production, this would integrate with payment gateway (Stripe, PayPal, etc.)
      const paymentResult = await this.simulatePaymentGateway(payment);

      if (paymentResult.success) {
        payment.status = 'COMPLETED';
        payment.transactionId = paymentResult.transactionId;
        payment.processedAt = new Date();
      } else {
        payment.status = 'FAILED';
        payment.failureReason = paymentResult.reason;
      }

      await payment.save();

      return {
        success: paymentResult.success,
        paymentId: payment._id,
        transactionId: payment.transactionId,
        status: payment.status,
        message: paymentResult.success ? 'Payment processed successfully' : paymentResult.reason
      };
    } catch (error) {
      throw error;
    }
  }

  // Simulate payment gateway processing
  async simulatePaymentGateway(payment) {
    return new Promise((resolve) => {
      setTimeout(() => {
        // 90% success rate for simulation
        const success = Math.random() > 0.1;
        
        if (success) {
          resolve({
            success: true,
            transactionId: `TXN-${uuidv4()}`
          });
        } else {
          const reasons = [
            'Insufficient funds',
            'Card declined',
            'Payment gateway timeout',
            'Invalid card details'
          ];
          resolve({
            success: false,
            reason: reasons[Math.floor(Math.random() * reasons.length)]
          });
        }
      }, 1000); // Simulate processing delay
    });
  }

  // Get payment by ID
  async getPaymentById(paymentId) {
    try {
      const payment = await Payment.findById(paymentId);
      if (!payment) {
        throw new Error('Payment not found');
      }
      return payment;
    } catch (error) {
      throw error;
    }
  }

  // Get payment by order ID
  async getPaymentByOrderId(orderId) {
    try {
      const payment = await Payment.findOne({ orderId });
      if (!payment) {
        throw new Error('Payment not found for this order');
      }
      return payment;
    } catch (error) {
      throw error;
    }
  }

  // Get all payments
  async getAllPayments() {
    try {
      return await Payment.find().sort({ createdAt: -1 });
    } catch (error) {
      throw error;
    }
  }

  // Refund payment
  async refundPayment(paymentId) {
    try {
      const payment = await Payment.findById(paymentId);
      
      if (!payment) {
        throw new Error('Payment not found');
      }

      if (payment.status !== 'COMPLETED') {
        throw new Error('Only completed payments can be refunded');
      }

      payment.status = 'REFUNDED';
      payment.updatedAt = new Date();
      await payment.save();

      return {
        success: true,
        message: 'Payment refunded successfully',
        paymentId: payment._id
      };
    } catch (error) {
      throw error;
    }
  }

  // Get payment status (for inter-service communication)
  async getPaymentStatus(orderId) {
    try {
      const payment = await Payment.findOne({ orderId });
      
      if (!payment) {
        return {
          exists: false,
          status: null
        };
      }

      return {
        exists: true,
        status: payment.status,
        paymentId: payment._id,
        transactionId: payment.transactionId,
        amount: payment.amount
      };
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new PaymentService();
